neps/nep-0330.md

```diff
@@ -31,23 +35,36 @@ There is a lot of information that can be held about a contract. Ultimately, we
 
 Successful implementations of this standard will introduce a new  (`ContractSourceMetadata`) struct that will hold all the necessary information to be queried for. This struct will be kept on the contract level.
 
-The metadata will include three optional fields:
+The metadata will include optional fields:
 
-- `version`: a string that references the specific commit hash or version of the code that is currently deployed on-chain. This can be included regardless of whether or not the contract is open-sourced and can also be used for organizational purposes.
+- `version`: a string that references the specific commit hash or version of the code currently deployed on-chain. If the link is to a GitHub repo URL, this is crucial, as it is needed to make the contract reproducible.
 - `link`: a string that references the link to the open-source code. This can be anything such as Github or a CID to somewhere on IPFS.
 - `standards`: a list of objects (see type definition below) that enumerates the NEPs supported by the contract. If this extension is supported, it is advised to also include NEP-330 version 1.1.0 in the list (`{standard: "nep330", version: "1.1.0"}`).
+- `build_details`: a build details object (see type definition below) that contains all the necessary information about how the contract was built, making it possible for others to reproduce the same WASM of this contract.
 
 ```ts
 type ContractSourceMetadata = {
   version: string|null, // optional, commit hash being used for the currently deployed wasm. If the contract is not open-sourced, this could also be a numbering system for internal organization / tracking such as "1.0.0" and "2.1.0".
   link: string|null, // optional, link to open source code such as a Github repository or a CID to somewhere on IPFS.
   standards: Standard[]|null, // optional, standards and extensions implemented in the currently deployed wasm e.g. [{standard: "nep330", version: "1.1.0"},{standard: "nep141", version: "1.0.0"}].
+  build_details: BuildDetails|null, // optional, details that are required for contract WASM reproducibility.
 }
 
 type Standard {
     standard: string, // standard name e.g. "nep141"
     version: string, // semantic version number of the Standard e.g. "1.0.0"
 }
+
+type BuildDetails {
+	env_image: EnvImage, // reference to a reproducible build environment image, e.g. "{tag: "sourcescan/cargo-near:0.6.0", digest: "sha256:bf488476d9c4e49e36862bbdef2c595f88d34a295fd551cc65dc291553849471"}"
+	contract_path: string|null // relative path to contract folder within the source code e.g. "src/contract". Often, it is the root of the repository, so can be omitted. 
+	build_command: string[] // the exact command that was used to build the contract, with all the flags e.g. ["cargo", "near", "build", "no-abi"].
+}
+
+type EnvImage {
```

Comment:
```suggestion
type Container {
```

Reply:
@robert-zaremba A container is a running instance of the image, so I believe it's still an image.

Reply:
Originally, it was named `build_environment_ref`, and it was expected to be either `docker.io/frol/near-builder@hash-of-the-image` or something less conventional (torrent link, IPFS cid, etc)

Reply:
```ts
type BuildInfo {
    build_environment_ref: string, // reference to a reproducible build environment image, e.g. "sourcescan/cargo-near@sha256:bf488476d9c4e49e36862bbdef2c595f88d34a295fd551cc65dc291553849471" or something else pointing to the build environment.
```

Now using `build_environment_ref` in 18968ea09c965845b92ab7fdb39a8a1df319b7cd

Reply:
honestly, `Env` in `EnvImage` doesn't sound right. `build_environment_ref` is even more cryptic. Why not `BuilderImage` . We can also say `DockerImage` if we don't want to use _Container_

Reply:
I like the name 'builder_image', but I think the original idea was to make it more abstract. This way, the reference could be something other than just a Docker image, for example, a link to a CD pipeline that creates the release.